# Import Packages

In [1]:
#Load Packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os as os
import requests
import io
import urllib.request
from sklearn.metrics import mean_squared_error
from scipy.io import arff
%matplotlib inline 

In [2]:
%pwd

'C:\\Users\\JM025575\\Predictive Models Class\\Week 3\\code'

In [3]:
cd /Users/JM025575/Predictive Models Class/data

C:\Users\JM025575\Predictive Models Class\data


# Load Data

In [4]:
prostate = pd.read_csv("prostate.csv", sep=",")
#print type of object for target
print("Data type", prostate.lpsa.dtype)
#Dimensions of dataset
print("Shape of Data", prostate.shape)
#Colum names
print("Colums Names", prostate.columns)
#See top few rows of dataset
prostate.head(10)

Data type float64
Shape of Data (97, 9)
Colums Names Index(['lcavol', 'lweight', 'age', 'lbph', 'svi', 'lcp', 'gleason', 'pgg45',
       'lpsa'],
      dtype='object')


lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45      lpsa
0 -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0 -0.430783
1 -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0 -0.162519
2 -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20 -0.162519
3 -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0 -0.162519
4  0.751416  3.432373   62 -1.386294    0 -1.386294        6      0  0.371564
5 -1.049822  3.228826   50 -1.386294    0 -1.386294        6      0  0.765468
6  0.737164  3.473518   64  0.615186    0 -1.386294        6      0  0.765468
7  0.693147  3.539509   58  1.536867    0 -1.386294        6      0  0.854415
8 -0.776529  3.539509   47 -1.386294    0 -1.386294        6      0  1.047319
9  0.223144  3.244544   63 -1.386294    0 -1.386294        6      0  1.047319

In [5]:
prostate.columns.tolist()

['lcavol', 'lweight', 'age', 'lbph', 'svi', 'lcp', 'gleason', 'pgg45', 'lpsa']

In [6]:
prostate.shape

(97, 9)

In [7]:
# Looking at the descriptive statistics of the target variable
prostate.lpsa.describe()

count    97.000000
mean      2.478387
std       1.154329
min      -0.430783
25%       1.731656
50%       2.591516
75%       3.056357
max       5.582932
Name: lpsa, dtype: float64

## Data Pre-Processing

In [8]:
#Define the features and target, lpsa
prostate.target=prostate.lpsa
prostate.features=prostate.drop(['lpsa'], axis=1)                         
print(prostate.target.shape)
print(prostate.features.shape)
prostate.head()

(97,)
(97, 8)


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  
C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  This is separate from the ipykernel package so we can avoid doing imports until


lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45      lpsa
0 -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0 -0.430783
1 -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0 -0.162519
2 -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20 -0.162519
3 -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0 -0.162519
4  0.751416  3.432373   62 -1.386294    0 -1.386294        6      0  0.371564

# Ridge Regression

In [9]:
# Ridge Regression package
from sklearn.linear_model import Ridge

## Basic Ridge Model

In [10]:
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 1.0)
model_RG.fit(prostate.features, prostate.target)
print(model_RG)
# make predictions
expected_RG = prostate.target
predicted_RG= model_RG.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.7682943732632037 [ 0.58397689  0.43665382 -0.01896769  0.10638469  0.68842174 -0.08707084
  0.03692423  0.00465442]
MSE 0.45596067462452833


The basic Ridge Regression model has a MSE of 0.4559 which I will work to lower in my following model adjustments.

In [11]:
#pd.DataFrame(np.transpose(model_RG.coef_))

In [12]:
#Adjusting the alpha in my model from the default of 1 to a 5.
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 5.0)
model_RG.fit(prostate.features, prostate.target)
print(model_RG)
# make predictions
expected_RG = prostate.target
predicted_RG= model_RG.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=5.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 1.0375988316496547 [ 0.56679975  0.37688704 -0.01683074  0.10564496  0.49540035 -0.03622384
  0.0192159   0.00491179]
MSE 0.4644668982849343


My second Ridge Regression model I adjusted my alpha, or error, up to 5.0.  This caused my MSE to go up slightly to 0.46446.

In [13]:
#Adjusting the alpha in my model from 5 to 0.5.
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 0.5)
model_RG.fit(prostate.features, prostate.target)
print(model_RG)
# make predictions
expected_RG = prostate.target
predicted_RG= model_RG.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=0.5, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.7210733869670383 [ 0.58560575  0.44540971 -0.01929328  0.10667301  0.72504385 -0.09585295
  0.04073685  0.00459521]
MSE 0.45547461120347865


My third Ridge Regression model, I adjusted my alpha down to 0.5 to see if that would cause my MSE to go back down.  It did cause it to go down although it is just slightly lower than the basic model at 0.4554.

## Grid Search to narrow down alpha and Cross Validation 10 times

In [15]:
# use a full grid over several parameters
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [0.01, 0.001, 0.0001, 0.05, 0.005, 0.75, 0.0005, 0.9]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_RG, param_grid=param_grid,n_jobs=-1, cv = 10)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.9}


Using a grid search, I ran 8 different alpha values through my model.  I use the grid search because it is a more efficient way of trying to find what my most ideal alpha value is rather than plugging one in at a time and seeing how my MSE adjusts.  I also used cross validation with the grid search cross validating 10 times.  Cross validation allows my model to be run a specific number of times, 10 in this case, each time taking different samples of data to run the model against for my various alphas I am using in the grid search.  Cross validation is a good way to see if we are overfitting our model. My result was that my best alpha is 0.9 demonstrating that continuing to lower my alpha value is not always the best way to get the lowest MSE.

## Ridge Model with Best Alpha From Gridsearch

In [16]:
#Ridge model with alpha of 0.9
# fit a ridge regression model to the data
model_RG = Ridge(alpha = 0.9)
model_RG.fit(prostate.features, prostate.target)
print(model_RG)
# make predictions
expected_RG = prostate.target
predicted_RG= model_RG.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_RG.intercept_, model_RG.coef_)
print("MSE", mean_squared_error(expected_RG, predicted_RG))

Ridge(alpha=0.9, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Coef 0.7591778129829336 [ 0.58431758  0.43838111 -0.01903145  0.10643602  0.69542291 -0.08876642
  0.03764465  0.00464333]
MSE 0.4558435387636189


My final Ridge model I used the 0.9 alpha that was considered my best alpha based on the grid search and cross validation I did in the previous step.  This produced a MSE of 0.4558.  Based purely on the MSE, I would say that my best Ridge Regression model is my 3rd one with the alpha of 0.5.

# LASSO Regression

In [17]:
# Lasso Regression-Least Absolute Shrinkage and Selection Operator
from sklearn.linear_model import Lasso

## Basic LASSO Model

In [18]:
# fit a LASSO model to the data
model_LAS = Lasso(alpha=1)
model_LAS.fit(prostate.features, prostate.target)
print(model_LAS)
# make predictions
expected_LAS = prostate.target
predicted_LAS = model_LAS.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 2.0879365603204247 [0.         0.         0.         0.         0.         0.
 0.         0.01601424]
MSE 1.0848110162374005


The basic Lasso Regression model produced a MSE of 1.0848, with the alpha set to 1.

In [19]:
# Adjust the alpha to 0.01 from 1.0
model_LAS = Lasso(alpha=0.01)
model_LAS.fit(prostate.features, prostate.target)
print(model_LAS)
# make predictions
expected_LAS = prostate.target
predicted_LAS = model_LAS.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=0.01, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.0301298227422002 [ 0.57839375  0.4109548  -0.0174482   0.10309292  0.6346103  -0.06304874
  0.          0.00498874]
MSE 0.45793415168104235


For my second Lasso model, I adjusted my alpha down to 0.01 which caused my MSE to drop all the way down to 0.4579.

## Full Grid Search With 3 Different Grids

### Grid 1

In [20]:
param_grid = {"alpha": [0.05, 0.01, 0.005]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.05}


For my first grid, I am feeding my Lasso model 3 alphas, including the alpha I used in my previous model that saw my MSE drop from the original model.  I also gave it an alpha that is above and below that value to see for my second grid where I need to adjust to next.  For grid one, my best alpha was 0.05.

### Grid 2

In [21]:
param_grid = {"alpha": [0.05, 0.005, 0.0005]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.05}


For my second grid, I again fed it 3 values including 0.05 from my last model.  This time, the grid search produced a result saying my best alpha is still 0.05.

### Grid 3

In [24]:
param_grid = {"alpha": [0.05, 0.1, 0.5]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_LAS, param_grid=param_grid,n_jobs=-1)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.1}


After running the grid search 3 different times, adjusting the alphas back up this time, I believe that using an alpha of 0.1 will be best for my model.

## Final Lasso Model With Best Alpha

In [23]:
# Adjust the alpha to 0.1 as indicated from the grid search
model_LAS = Lasso(alpha=0.1)
model_LAS.fit(prostate.features, prostate.target)
print(model_LAS)
# make predictions
expected_LAS = prostate.target
predicted_LAS = model_LAS.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_LAS.intercept_,model_LAS.coef_)
print("MSE", mean_squared_error(expected_LAS, predicted_LAS))

Lasso(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.6699333664535905 [ 0.57700424  0.06179715 -0.00577247  0.07308461  0.          0.
 -0.          0.00677141]
MSE 0.5576559643302158


For my final Lasso model, I used my best alpha of 0.1 that was produced from my 3 separate grid searches and narrowing down my results.  For this model, that alpha value gives an output of a MSE of 0.5576 which is not our best result from the Lasso model, and not as good as our best results from the Ridge models. My best Lasso model had an alpha of 0.01.

# ElasticNet Regression

In [25]:
# ElasticNet Regression
from sklearn.linear_model import ElasticNet

In [26]:
# Basic ElasticNet Regression Model
model_EN = ElasticNet(alpha=1)
model_EN.fit(prostate.features, prostate.target)
print(model_EN)
# make predictions
expected_EN = prostate.target
predicted_EN = model_EN.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=1, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.9313322879124133 [0.15562453 0.         0.         0.         0.         0.
 0.         0.01382035]
MSE 0.8798461494497223


My first ElasticNet model used the default values, which means an alpha of 1.  This produced an MSE of 0.8798.

In [27]:
# ElasticNet Regression Model adjusting Alpha from 1 to 0.5
model_EN = ElasticNet(alpha=0.5)
model_EN.fit(prostate.features, prostate.target)
print(model_EN)
# make predictions
expected_EN = prostate.target
predicted_EN = model_EN.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.5, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.734484615564524 [0.36370419 0.         0.         0.         0.         0.
 0.         0.01037257]
MSE 0.6912923472841728


My second ElasticNet model adjusted the alpha down to 0.5 which produced a lower MSE of 0.6912.

In [28]:
# ElasticNet Regression Model adjusting Alpha from 0.5 to 0.05
model_EN = ElasticNet(alpha=0.05)
model_EN.fit(prostate.features, prostate.target)
print(model_EN)
# make predictions
expected_EN = prostate.target
predicted_EN = model_EN.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.05, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.2225577086019637 [ 0.55774379  0.32375852 -0.01396712  0.09627958  0.38238527 -0.
  0.          0.00491596]
MSE 0.4755112050977968


My 3rd model used an alpha of 0.05, which finally saw my MSE go down to 0.4755.  This did lower the MSE, but I will next use a grid search combined with cross validation to see if continued dropping of the alpha can bring my MSE even lower.

## Grid Search to narrow down alpha with Cross Validation 5 Times

In [29]:
# use a full grid over several parameters
param_grid = {"alpha": [0.00005, 0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25]}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_EN, param_grid=param_grid,n_jobs=-1, cv = 5)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'alpha': 0.25}


Grid search between 9 alpha values and a cross validation being performed 5 times yields a best alpha result of 0.25.  My grid search included values below and above the previously used alpha of 0.05.  The cross validation allowed each of those alphas to run through the model 5 times using separate samples of data.  The combination of the alphas in the grid search and the cross validation produced a best alpha of 0.25.

In [30]:
# ElasticNet Regression Model adjusting Alpha to 0.0001 as indicated as the best alpha for the model
model_EN = ElasticNet(alpha=0.25)
model_EN.fit(prostate.features, prostate.target)
print(model_EN)
# make predictions
expected_EN = prostate.target
predicted_EN = model_EN.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_EN.intercept_, model_EN.coef_)
print("MSE", mean_squared_error(expected_EN, predicted_EN))

ElasticNet(alpha=0.25, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Coef 1.707335291612384 [ 0.49643943  0.         -0.0014447   0.05931984  0.          0.00820404
 -0.          0.00773701]
MSE 0.5967985948322396


My final version of the ElasticNet model used my best alpha from above, 0.25.  This produced a MSE for my ElasticNet Regressions of 0.5967 which is not my best MSE for the ElasticNet telling me there would be additional parameters that need adjusting to continue further lowering the MSE for my model.

# SGD Regressor

In [31]:
# SGD Regressor
from sklearn.linear_model import SGDRegressor

In [32]:
# Basic SGD Regressor Model, l2 penalty 
model_SGD = SGDRegressor()
model_SGD.fit(prostate.features, prostate.target)
print(model_SGD)
# make predictions
expected_SGD = prostate.target
predicted_SGD = model_SGD.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=0.0001, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l2',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [-6.77929694e+08] [-7.70009280e+09  4.58271754e+10  8.92255776e+10 -8.27413691e+10
 -2.82687433e+10 -2.41121970e+10 -3.09273183e+10  1.00083452e+11]
MSE 7.421110338012683e+25


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


My first SGD Regressor model using all of the default parameter values, produced an MSE of 7.4211.

In [33]:
# SGD Regressor Model, adjusting penalty to l1 and alpha to 0.01
model_SGD = SGDRegressor(alpha = 0.01, penalty = 'l1')
model_SGD.fit(prostate.features, prostate.target)
print(model_SGD)
# make predictions
expected_SGD = prostate.target
predicted_SGD = model_SGD.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=0.01, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l1',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [6.30808785e+09] [ 3.38186965e+10 -4.48634201e+09 -2.58242536e+10 -1.36546973e+11
  5.39619927e+09 -4.24063811e+10  7.80686344e+10 -1.72911269e+11]
MSE 5.224196215581104e+25


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


For my second model, I adjusted my alpha up to 0.01 from the default of 0.0001 and changed my penalty from the default L2 to L1. These changes caused my MSE to go down from the basic model to 5.2241.

In [35]:
# SGD Regressor Model, adjusting penalty to l1 and alpha to 0.00001
model_SGD = SGDRegressor(alpha = 0.00001, penalty = 'l2')
model_SGD.fit(prostate.features, prostate.target)
print(model_SGD)
# make predictions
expected_SGD = prostate.target
predicted_SGD = model_SGD.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=1e-05, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='l2',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [-2.25924964e+09] [ 7.45251104e+10 -1.98566627e+10  9.36716234e+10 -1.22394548e+11
 -7.50006899e+09 -2.99797502e+10 -5.90037277e+08 -6.86371648e+10]
MSE 2.21168628916299e+25


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


My third model I set my penalty back to at L2, and adjusted the alpha below the default value to 0.00001.  This lowered my MSE to 2.2116, which was my best SGD model yet.

## Grid Search to narrow down alpha with Cross Validation 10 Times

In [36]:
# use a full grid over several parameters
param_grid = {"penalty": ['none', 'l2', 'l1', 'elasticnet']}  #different penalties I am playing with

# run grid search
grid_search = GridSearchCV(model_SGD, param_grid=param_grid,n_jobs=-1, cv = 10)
grid_search.fit(prostate.features, prostate.target)
print("Best", grid_search.best_params_) 

Best {'penalty': 'none'}


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Next, I used a grid search using the 4 possible penalty values and a cross validation ten times to see which of the penalty values is best to use.  My result was that the best penalty to use is 'none'.

In [38]:
# SGD Regressor Model, adjusting penalty 'none' based on the grid search results and cross validation
model_SGD = SGDRegressor(alpha = 0.00001, penalty = 'none')
model_SGD.fit(prostate.features, prostate.target)
print(model_SGD)
# make predictions
expected_SGD = prostate.target
predicted_SGD = model_SGD.predict(prostate.features)
# summarize the fit of the model
print("Coef", model_SGD.intercept_, model_SGD.coef_)
print("MSE", mean_squared_error(expected_SGD, predicted_SGD))

SGDRegressor(alpha=1e-05, average=False, epsilon=0.1, eta0=0.01,
       fit_intercept=True, l1_ratio=0.15, learning_rate='invscaling',
       loss='squared_loss', max_iter=None, n_iter=None, penalty='none',
       power_t=0.25, random_state=None, shuffle=True, tol=None, verbose=0,
       warm_start=False)
Coef [-9.04905494e+08] [-2.38270854e+09  2.25756649e+09 -6.29617672e+10 -2.07780326e+10
  2.20871857e+10  1.21835735e+11  2.52941673e+09 -2.21714819e+11]
MSE 1.2800948445792925e+26


C:\Users\JM025575\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


Finally, I used the result from my grid search and cross validation and paired that with the lowest alpha I had used and ended up getting my lowest MSE yet with this model, 1.28009.

# Hold for explanation of what best model is and why

After reviewing my 4 different Ridge, Lasso, ElasticNet, and SGD Regressor, the best model I would say that my best model was my third Ridge Regression model, where I used an alpha of 0.5 and got a MSE of 0.4554.  I got various other MSE values that came close to that after tuning those models in various ways and using grid search and cross validation to more efficiently search for my best alpha and penalty values.  However, based solely on the lowest MSE, my Ridge Regression Model with the alpha of 0.5 is my best performing model.